drivers/timer/apic_timer.c

```diff
@@ -0,0 +1,378 @@
+ * the LOAPIC_TIMER register, so be sure to account for that in any changes.
+ */
+
+int z_clock_driver_init(struct device *device)
+{
+	u32_t val;
+
+	/* 1. set DCR - clock divider */
+
+	val = x86_read_loapic(LOAPIC_TIMER_CONFIG);
+	val &= ~DCR_DIVIDER_MASK;
+	val |= DCR_DIVIDER;
+	x86_write_loapic(LOAPIC_TIMER_CONFIG, val);
+
+	/* 2. set LVT - timer mode */
+
+	val = x86_read_loapic(LOAPIC_TIMER);
+	val &= ~LVT_MODE_MASK;
+	val |= LVT_MODE;
+	x86_write_loapic(LOAPIC_TIMER, val);
+
+	/* 3. wire up interrupt (this will configure remaining LVT bits) */
+
+	IRQ_CONNECT(CONFIG_APIC_TIMER_IRQ,
+		CONFIG_APIC_TIMER_IRQ_PRIORITY,
+		isr, 0, 0);
+	irq_enable(CONFIG_APIC_TIMER_IRQ);
+
+	/* 4. schedule the first interrupt */
+
+	x86_write_loapic(LOAPIC_TIMER_ICR, CYCLES_PER_TICK);
+
+	return 0;
+}
+
+#ifdef CONFIG_TICKLESS_KERNEL
+
+/*
+ * Based on my understanding of the contract described in
+ * include/drivers/system_timer.h, this function acts as follows:
+ *
+ * 1. n <= 0
+ *
+ *     These are treated the same as a call to n == 1, that
+ *     is, announce at the next tick boundary (see 3).
+ *
+ * 2. (a) n == K_FOREVER or
+ *    (b) n > MAX_TICKS (which includes n == INT_MAX)
+ *
+ *     In these cases we sleep for MAX_TICKS. Note that we could
+ *     completely shut down the clock in case (a) but we don't do
+ *     any power management so we don't distinguish at the moment.
+ *
+ * 3. MAX_TICKS >= n >= 1
+ *
+ *     Arrange for an announcement at the 'n'th tick boundary
+ *     in the future. Note that this means that if n == 1, we
+ *     could announce almost immediately.
+ */
+
+void z_clock_set_timeout(s32_t n, bool idle)
+{
+	ARG_UNUSED(idle);
+
+	u32_t icr;
+	u32_t ccr;
+
+	/*
+	 * We split the calculation of the correct ICR into two
+	 * parts, full_cycles and partial_cycles. The sum of these
+	 * two will result in an ICR value that will expire the
+	 * timer at the next tick boundary.
+	 */
+
+	int   full_ticks;	/* number of FULL ticks we'll wait */
+	u32_t full_cycles;	/* full_ticks represented as cycles */
+	u32_t partial_cycles;	/* number of cycles in the partial tick */
+
+	if (n < 1) {
+		full_ticks = 0;
+	} else if ((n == K_FOREVER) || (n > MAX_TICKS)) {
+		full_ticks = MAX_TICKS - 1;
+	} else {
+		full_ticks = n - 1;
+	}
+
+	full_cycles = full_ticks * CYCLES_PER_TICK;
+
+	/*
+	 * There's a wee race condition here. The timer may expire while
+	 * we're busy reprogramming it; an interrupt will be queued at the
+	 * local APIC and the ISR will be called too early, roughly right
+	 * after we unlock, and not because the count we just programmed has
+	 * counted down. Luckily this situation is easy to detect, which is
+	 * why the ISR actually checks to be sure the CCR is 0 before acting.
+	 */
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	icr = x86_read_loapic(LOAPIC_TIMER_ICR);
+	ccr = x86_read_loapic(LOAPIC_TIMER_CCR);
+	total_cycles += (icr - ccr);
+	partial_cycles = CYCLES_PER_TICK - (total_cycles % CYCLES_PER_TICK);
```

Comment:
FWIW: the complexity involved in calculating "partial cycles" goes away entirely if you use an Invariant TSC as your time base (along the potential loss of precision when interrupts are late).  Just calculate your "desired TSC expiration" based on the last announcement time, subtract from the current value, and that's your new ICR.

Reply:
The driver is written to be useful on systems without a suitable TSC... I don't know if it's worth starting the special-case spaghetti here, but point taken.  If a usable TSC is assumed, the driver could certainly look a lot different.

Reply:
In fact, on that note, I'll assign myself a github to-do to write a TSC-only driver, then the TSC stuff can come out of here entirely. The idea in this driver is to assume the least common denominator, with a tiny concession to the existence of TSCs where the cost/benefit was high.